src/components/organisms/LocalReplace.tsx

```diff
@@ -0,0 +1,448 @@
+  onUpdateSearchOptions: (searchOptions: Partial<SearchReplaceOptions>) => void
+  onReplaceToggle?: (nextState?: boolean) => void
+  onReplaceQueryChange?: (newSearchQuery: string) => void
+  onReplacementFinished?: () => void
+  onFocusSearchInput?: () => void
+  onReplaceClose?: () => void
+}
+
+const LocalReplace = ({
+  codeMirror,
+  replaceQuery,
+  searchOptions,
+  numberOfFoundItems,
+  focusingReplace,
+  navigateToNext,
+  onReplaceToggle,
+  onReplaceQueryChange,
+  onReplacementFinished,
+  onFocusSearchInput,
+  onReplaceClose,
+  onUpdateSearchOptions,
+}: LocalReplaceProps) => {
+  const replaceTextAreaRef = useRef<HTMLTextAreaElement>(null)
+  const [replaceValue, setReplaceValue] = useState(replaceQuery)
+  const [preservingCaseReplace, setPreservingCaseReplace] = useState<boolean>(
+    searchOptions.preservingCaseReplace
+  )
+
+  const getNumberOfTextAreaRows = useMemo(() => {
+    const replaceNumLines = replaceValue ? replaceValue.split('\n').length : 0
+    return replaceNumLines == 0 || replaceNumLines == 1
+      ? 1
+      : replaceNumLines + 1
+  }, [replaceValue])
+
+  const updateReplaceValue: ChangeEventHandler<HTMLTextAreaElement> = useCallback(
+    (event) => {
+      setReplaceValue(event.target.value)
+      if (onReplaceQueryChange != null) {
+        onReplaceQueryChange(event.target.value)
+      }
+    },
+    [onReplaceQueryChange]
+  )
+
+  const handleOnReplaceClose = useCallback(() => {
+    if (onReplaceClose != null) {
+      onReplaceClose()
+    }
+  }, [onReplaceClose])
+
+  const focusReplaceTextAreaInput = useCallback((replaceValueLength = 0) => {
+    if (replaceTextAreaRef.current == null) {
+      return
+    }
+    replaceTextAreaRef.current.focus()
+    if (replaceValueLength > 0) {
+      replaceTextAreaRef.current.selectionEnd += replaceValueLength
+      replaceTextAreaRef.current.selectionStart = replaceValueLength
+    }
+  }, [])
+
+  const addNewlineToReplaceValue = useCallback(() => {
+    if (replaceTextAreaRef.current === null) {
+      return
+    }
+    const startPosition = replaceTextAreaRef.current.selectionStart
+    const endPosition = replaceTextAreaRef.current.selectionEnd
+    if (startPosition == endPosition) {
+      // no selection -> add newline at this location
+      replaceTextAreaRef.current.setRangeText(
+        '\n',
+        startPosition,
+        endPosition,
+        'end'
+      )
+    } else {
+      // selection -> replace it with newline
+      replaceTextAreaRef.current.setRangeText(
+        '\n',
+        startPosition,
+        endPosition,
+        'end'
+      )
+    }
+    setReplaceValue(replaceTextAreaRef.current.value)
+    if (onReplaceQueryChange != null) {
+      onReplaceQueryChange(replaceValue)
+    }
+    focusReplaceTextAreaInput()
+  }, [focusReplaceTextAreaInput, onReplaceQueryChange, replaceValue])
+
+  const replaceSingleMarker = useCallback(
+    (marker: TextMarker) => {
+      const markerPosition:
```

Comment:
Not sure how to refactor this to not use some type assertion. The new codemirror typings have MarkerRange and Position union type for item returned by `marker.find()`